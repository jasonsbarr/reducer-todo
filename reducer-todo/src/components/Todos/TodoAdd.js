import React from "react";

const TodoAdd = {
  return <div>Todo Form</div>
};

export default TodoAdd;
