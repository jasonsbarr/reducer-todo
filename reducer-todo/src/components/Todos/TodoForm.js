import React, { useState } from "react";
import { useLocation } from "react-router-dom";

const defaultValues = {
  item: "",
  due_at_date: "",
  due_at_time: "",
  priority: 4,
};

const TodoForm = ({ handleSubmit, todo = undefined }) => {
  const { pathname } = useLocation();

  const formValues = {
    item: todo ? todo.item : defaultValues.item,
    due_at_date:
      todo && todo.due_at
        ? todo.due_at.toISOString().split("T")[0]
        : defaultValues.due_at_date,
    due_at_time:
      todo && todo.due_at
        ? todo.due_at
            .toISOString()
            .split("T")[1]
            .split(".")[0]
        : defaultValues.due_at_time,
    priority: todo ? todo.priority : defaultValues.priority,
  };

  const [formState, setFormState] = useState(formValues);

  const handleChange = event => {
    setFormState({
      ...formState,
      [event.target.name]: event.target.value,
    });
  };

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="item">
        Todo item (required):
        <input
          type="text"
          name="item"
          id="item"
          value={formState.item}
          placeholder="Todo item goes here..."
          onChange={handleChange}
        />
      </label>
      <label htmlFor="due_at">
        Due date:
        <span id="due_at">
          <input
            type="date"
            name="due_at_date"
            id="due_at_date"
            value={formState.due_at_date}
            onChange={handleChange}
          />
          <input
            type="time"
            name="due_at_time"
            id="due_at_time"
            value={formState.due_at_time}
            onChange={handleChange}
          />
        </span>
      </label>
      <label htmlFor="priority">
        Priority:
        <select
          name="priority"
          id="priority"
          value={formState.priority}
          selected={4}
          onChange={handleChange}
        >
          <option value={1}>1</option>
          <option value={2}>2</option>
          <option value={3}>3</option>
          <option value={4}>4</option>
        </select>
      </label>
      <button type="submit">
        {pathname === "/todo/add" ? "Add Todo" : "Edit Todo"}
      </button>
    </form>
  );
};

export default TodoForm;
