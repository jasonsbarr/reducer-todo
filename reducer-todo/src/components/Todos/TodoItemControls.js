import React from "react";
import { useHistory } from "react-router-dom";
import Axios from "axios";
import { DELETE_TODO, EDIT_TODO } from "../../actions/todos";
import useTodos from "../../hooks/useTodos";

const TodoItemControls = ({ todo }) => {
  const apiUrl = `http://localhost:4000/todos`;
  const history = useHistory();
  const [, dispatch] = useTodos();

  const handleCompleteTodo = () => {
    const changed = {
      ...todo,
      completed: !todo.completed,
      updated_at: new Date(Date.now()),
    };

    const options = {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      data: JSON.stringify(changed),
      url: `${apiUrl}/${todo.id}`,
    };

    dispatch({ type: EDIT_TODO, payload: changed });

    Axios(options).catch(err => console.error(err));
  };

  const handleEditTodo = () => {
    history.push(`/todo/${todo.id}/edit`);
  };

  const handleDeleteTodo = () => {
    dispatch({ type: DELETE_TODO, payload: todo.id });

    const options = {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      url: `${apiUrl}/${todo.id}`,
    };

    Axios(options).catch(err => console.error(err));
  };

  return (
    <div className="todo-item-controls">
      <button onClick={handleCompleteTodo}>
        {todo.completed ? "⍻" : "✓"}
      </button>
      <button onClick={handleEditTodo}>🖉</button>
      <button id={`delete-${todo.id}`} onClick={handleDeleteTodo}>
        ×
      </button>
    </div>
  );
};

export default TodoItemControls;
