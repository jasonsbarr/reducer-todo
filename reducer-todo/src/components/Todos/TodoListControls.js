import React from "react";

const TodoListControls = ({ onClearCompleted }) => (
  <button onClick={onClearCompleted}>Clear Completed</button>
);

export default TodoListControls;
