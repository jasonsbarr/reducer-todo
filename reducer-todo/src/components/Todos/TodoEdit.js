import React, { useState } from "react";
import { useParams, useHistory } from "react-router-dom";
import useFetchSingleTodo from "../../hooks/useFetchSingleTodo";
import useTodos from "../../hooks/useTodos";
import TodoForm from "./TodoForm";
import { EDIT_TODO } from "../../actions/todos";
import Axios from "axios";

const TodoEdit = () => {
  const url = `http://localhost:4000/todos`;
  const { todoId } = useParams();
  const [, dispatch] = useTodos();
  const history = useHistory();
  const [todo, setTodo] = useState(null);
  const [message, setMessage] = useState("");

  useFetchSingleTodo(Number(todoId))
    .then(([todo, message]) => {
      setTodo(todo);
      setMessage(message);
    })
    .catch(err => {
      setMessage("Something's gone horribly wrong.");
      console.error(err);
    });

  const handleFormSubmit = event => {
    event.preventDefault();

    const editedTodo = {
      uuid: todo.uuid,
      item: event.target.item.value,
      priority: Number(event.target.priority.value),
      completed: todo.completed,
      due_at: new Date(
        `${event.target.due_at_date.value}T${event.target.due_at_time.value}Z`,
      ),
      created_at: todo.created_at,
      updated_at: new Date(Date.now()),
      id: todo.id,
    };

    dispatch({ type: EDIT_TODO, payload: editedTodo });

    const options = {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      data: JSON.stringify(editedTodo),
      url: `${url}/${todo.id}`,
    };

    Axios(options)
      .then(res => {
        history.push("/");
      })
      .catch(err => console.error(err));
  };

  return (
    <>
      <h1>Edit Todo</h1>
      <div className="form-container">
        {todo ? (
          <TodoForm todo={todo} handleSubmit={handleFormSubmit} />
        ) : (
          <div>{message}</div>
        )}
      </div>
    </>
  );
};

export default TodoEdit;
