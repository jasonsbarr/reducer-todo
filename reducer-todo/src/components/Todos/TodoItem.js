import React from "react";
import { Link } from "react-router-dom";
import TodoItemControls from "./TodoItemControls";
import returnRelativeDueDate from "../../utils/returnRelativeDueDate";

const TodoItem = ({
  todo,
  onCompleteTodo,
  onEditTodo,
  onDeleteTodo,
} = {}) => {
  const { id, item, priority, due_at, completed, uuid } = todo;

  return (
    <li
      id={`todo-${id}`}
      data-uuid={uuid}
      completed={completed ? 1 : 0}
      className={`todo-item${completed ? " completed" : ""}`}
    >
      <div className="todo-item-container">
        <p>
          <Link to={`/todo/${id}`}>{item}</Link>
        </p>
        <p>
          <span className="label">Priority:</span> {priority}
        </p>
        <p>{returnRelativeDueDate(due_at)}</p>
        <TodoItemControls todo={todo} />
      </div>
    </li>
  );
};

export default TodoItem;
