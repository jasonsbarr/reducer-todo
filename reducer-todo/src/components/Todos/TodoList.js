import React, { useState, useEffect } from "react";
import Axios from "axios";
import {
  SET_TODOS,
  TOGGLE_TODO_COMPLETE,
  CLEAR_COMPLETED_TODOS,
  CLEAR_ALL_TODOS,
  DELETE_TODO,
} from "../../actions/todos";
import useTodos from "../../hooks/useTodos";
import { TodoItem } from ".";
import TodoListControls from "./TodoListControls";

const TodoList = ({ todos }) => {
  const apiUrl = `http://localhost:4000/todos`;

  // Initialize reducer w/ initialState
  const [state, dispatch] = useTodos();
  const [message, setMessage] = useState("Loading...");

  const handleCompleteTodo = todo => {
    const changed = {
      ...todo,
      completed: !todo.completed,
      updated_at: new Date(Date.now()),
    };
    const options = {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      data: JSON.stringify(changed),
      url: `${apiUrl}/${todo.id}`,
    };

    dispatch({ type: TOGGLE_TODO_COMPLETE, payload: todo });

    Axios(options).catch(err => console.error(err));
  };

  const handleDeleteTodo = todo => {
    dispatch({ type: DELETE_TODO, payload: todo.id });

    const options = {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      url: `${apiUrl}/${todo.id}`,
    };

    Axios(options).catch(err => console.error(err));
  };

  const handleClearCompletedTodos = () =>
    dispatch({ type: CLEAR_COMPLETED_TODOS });

  // Initial fetch todos, dispatch to set state
  useEffect(() => {
    dispatch({ type: CLEAR_ALL_TODOS });
    Axios.get(apiUrl)
      .then(res =>
        dispatch({
          type: SET_TODOS,
          payload: res.data,
        }),
      )
      .catch(err =>
        setMessage("Something went wrong. Please try again later."),
      );
  }, [apiUrl, dispatch]);

  return (
    <>
      <h1>Todo List</h1>
      {state.todos && state.todos.length ? (
        <>
          <TodoListControls
            onClearCompleted={handleClearCompletedTodos}
          />
          <ul>
            {state.todos.map(todo => (
              <TodoItem
                onCompleteTodo={handleCompleteTodo}
                onDeleteTodo={handleDeleteTodo}
                key={todo.uuid}
                todo={todo}
              />
            ))}
          </ul>
        </>
      ) : (
        <div>{message}</div>
      )}
    </>
  );
};

export default TodoList;
