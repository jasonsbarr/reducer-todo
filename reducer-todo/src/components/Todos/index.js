import TodoItem from "./TodoItem";
import TodoList from "./TodoList";
import TodoAdd from "./TodoAdd";
import TodoSingle from "./TodoSingle";

export { TodoItem, TodoList, TodoAdd, TodoSingle };
