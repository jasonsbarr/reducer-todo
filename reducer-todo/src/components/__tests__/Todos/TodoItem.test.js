import React from "react";
import { render, cleanup, act } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { TodoItem } from "../../Todos";

afterEach(cleanup);

const todos = [
  {
    item: "Do something",
    completed: false,
    id: "todo-1",
  },
  {
    item: "Do something else",
    completed: true,
    id: "todo-2",
  },
];

describe("TodoItem component tests", () => {
  test("should render a todo item", () => {
    const { container } = render(
      <BrowserRouter>
        <TodoItem todo={todos[0]} />
      </BrowserRouter>,
    );

    const todoItem = container.firstChild;

    expect(todoItem.textContent).toMatch("Do something");
    expect(todoItem).toMatchSnapshot();
  });

  test("should have CSS class of completed if todo is completed", () => {
    const { container } = render(
      <BrowserRouter>
        <TodoItem todo={todos[1]} />
      </BrowserRouter>,
    );

    const todoItem = container.firstChild;

    expect(todoItem.textContent).toMatch("Do something else");
    expect(todoItem.classList.contains("completed")).toBe(true);
  });
});
