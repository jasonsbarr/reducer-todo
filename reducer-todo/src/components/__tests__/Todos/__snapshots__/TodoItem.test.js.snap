// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component tests should render a todo item 1`] = `
<li
  class="todo-item"
  completed="0"
  id="todo-todo-1"
>
  <div
    class="todo-item-container"
  >
    <p>
      <a
        href="/todo/todo-1"
      >
        Do something
      </a>
    </p>
    <button>
      âœ“
    </button>
    <button>
      ðŸ–‰
    </button>
    <button
      id="delete-todo-1"
    >
      Ã—
    </button>
  </div>
</li>
`;
